### PR TITLE
Marriage, Parentship and Lineage Agents for Reasoning Benchmark

### LINKED ISSUE #127: Agents for read performance with reasoning
## Description

Subsequent to the significant refactor this year, we must introduce new agents. We particularly want to focus on creating a variety of queries that demonstrate the different challenges in reasoner.

### PR DESCRIPTION
**Pull request by jmsfltchr**
## What is the goal of this PR?

We add new agents with the goal of benchmarking common reasoning queries. Two agents simply insert new data, and so are applicable for benchmarking on other databases. Reasoning queries do not have a fair comparison in other DBs and so are only implemented for TypeDB.

## What are the changes implemented in this PR?

New agents for TypeDB and Neo4j:
- `MarriageAgent`
- `ParenthoodAgent`

New reasoning agents for TypeDB only:
- `NationalityAgent` queries for people's nationality, determined based on their birthplace being contained in a country.
- `CitizenshipAgent` queries for citizens of a country. People can be citizens based on three rules. Either they are nationals, they have a parent as a national, or their partner in marriage is a national.
- `CoupleFriendshipAgent` queries for friendships. The name is to differentiate from `FriendshipAgent` (which inserts friendships), and also indicates the rule being benchmarked. The rule dictates that married couples are friends.
- `GrandparenthoodAgent` queries for `grandparenthood`, which is always inferred. Note this is not transitive, this is a simple two hops over `parenthood`.
- `LineageAgent` queries for `lineage`, which is a transitive relation over `parentship`.
- `MaritalStatusAgent` queries for an attribute on a person that shows their marital status based on whether they are in a `marriage`.

### CODE REVIEW

File: typedb/simulation.gql [hunk 1]
```diff
@@ -42,12 +44,17 @@ person sub entity,
     owns birth-date,
     owns gender,
     owns address,
+    owns marital-status,
     plays birth-place:child,
     plays residentship:resident,
+    plays nationality:national,
```
Comment (haikalpribadi) [inline typedb/simulation.gql:50]: `national` is a role for a `person`? are you sure? Dictionary says the meaning of `national` is:

> relating to or characteristic of a nation; common to a whole nation.

Doesn't sound like something a `person` would be, no?

File: typedb/simulation.gql [hunk 2]
```diff
@@ -84,7 +91,9 @@ region sub entity, abstract,
 
 continent sub region;
 country sub region,
-    owns currency;
+    owns currency,
+    plays nationality:state,
+    plays citizenship:state;
```
Comment (haikalpribadi) [inline typedb/simulation.gql]: Would `country` not be a better role than `state` here?

**Review by haikalpribadi — CHANGES_REQUESTED** [review]

File: typedb/simulation.gql (hunk 2 above)
Comment (jmsfltchr) [reply typedb/simulation.gql]: Yes let's do that

File: typedb/simulation.gql (hunk 1 above)
Comment (jmsfltchr) [reply typedb/simulation.gql:50]: That's the meaning when used as an adjective, used as a noun we have 

> a citizen of a particular country

To conform we'll also use `nation` rather than `state` as the other role in `nationality`.

File: Benchmark.java [hunk 3]
```diff
@@ -65,7 +65,7 @@ public static void main(String[] args) {
         return simulation;
     }
 
-    private static FactoryTracing initTracing(@Nullable Options.FactoryTracing options, String analysisName) {
+    private static FactoryTracing initTracing(@Nullable Options.FactoryTracing options, String databaseFullname) {
```
Comment (haikalpribadi) [inline Benchmark.java]: In the context of **this function**, you don't actually know anything about the fact that the `analysisName` is actually a `databaseName`. It just so happens, that's how it's used from the outside. But to name this variable based on how it's used from the outside would be a mistake, as you are leaking the knowledge on how the function is being used to define the logic of how the function is being written. The function and it's signature should be canonical and completely generic.

File: WORKSPACE [hunk 4]
```diff
@@ -94,6 +94,10 @@ rules_pkg_dependencies()
 load("@vaticle_bazel_distribution//github:deps.bzl", github_deps = "deps")
 github_deps()
 
+# Load //pip
+load("@vaticle_bazel_distribution//pip:deps.bzl", pip_deps = "deps")
+pip_deps()
```
Comment (haikalpribadi) [inline WORKSPACE:99]: What are we doing now that makes us need to add this? I can't see any new code that uses it.

File: common/concept/Parenthood.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package com.vaticle.typedb.benchmark.common.concept;
+
+public class Parenthood {
```
Comment (haikalpribadi) [inline common/concept/Parenthood.java:22]: How will the equality of the object instances of this class work without the equality and hash function? 🤔

File: common/concept/Marriage.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package com.vaticle.typedb.benchmark.common.concept;
+
+import java.time.LocalDateTime;
+
+public class Marriage {
```
Comment (haikalpribadi) [inline common/concept/Marriage.java:23]: How will the equality of the object instances of this class work without the equality and hash function? 🤔

File: common/params/Context.java [hunk 7]
```diff
@@ -71,31 +69,47 @@ public ExecutorService executor() {
...
+        iteration.incrementAndGet();
     }
 
     public LocalDateTime today() {
-        return LocalDateTime.of(LocalDate.ofYearDay(2000 + iteration.get(), 1), LocalTime.of(0, 0, 0));
+        return startDay().plusYears(iteration.get() - 1);
+    }
+
+    public LocalDateTime startDay() {
+        return LocalDateTime.of(LocalDate.ofYearDay(2000, 1), LocalTime.of(0, 0, 0));
```
Comment (haikalpribadi) [inline common/params/Context.java:112]: This should probably be parameterised and configurable, right?

File: simulation/agent/LineageAgent.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.List;
+
+public abstract class LineageAgent<TX extends Transaction> extends Agent<Country, TX> {
+
+    protected LineageAgent(Client<?, TX> client, Context context) {
+        super(client, context);
+    }
+
+    @Override
+    protected Class<? extends Agent> agentClass() {
```
Comment (haikalpribadi) [inline simulation/agent/LineageAgent.java:38]: Raw use of parameterised class `Agent`

File: simulation/agent/LineageAgent.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        try (TX tx = session.reasoningTransaction()) {
+            matchLineages(tx, country, context.startDay(), context.today());
+            if (context.isReporting()) {
+                throw new RuntimeException("Reports are not comparable for reasoning agents.");
+            }
+        }
+        return reports;
+    }
+
+    protected abstract void matchLineages(TX tx, Country country, LocalDateTime startDay, LocalDateTime today);
```
Comment (haikalpribadi) [inline simulation/agent/LineageAgent.java:59]: Why does matching lineages have to be parameterised by start and end date? 🤔

File: simulation/agent/MarriageAgent.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    @Override
+    protected List<Report> run(Session<TX> session, Country country, RandomSource random) {
+        List<Report> reports = new ArrayList<>();
+        try (TX tx = session.writeTransaction()) {
+            LocalDateTime partnerBirthDate = context.today().minusYears(context.ageOfAdulthood());
+            List<Person> women = matchPartner(tx, country, partnerBirthDate, Gender.FEMALE)
+                    .sorted(comparing(Person::email)).collect(toCollection(ArrayList::new));
+            List<Person> men = matchPartner(tx, country, partnerBirthDate, Gender.MALE)
+                    .sorted(comparing(Person::email)).collect(toCollection(ArrayList::new));
+            random.randomPairs(women, men).forEach(partners -> {
```
Comment (haikalpribadi) [inline simulation/agent/MarriageAgent.java]: Isn't the `partners` variable capturing only 1 pair of man and woman? Shouldn't it be `partner`?

File: simulation/agent/MarriageAgent.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+                                                             licence, context.today());
+                if (context.isReporting()) {
+                    assert inserted.isPresent();
+                    reports.add(new Report(list(partners.first().email(), partners.second().email(),
+                                                licence, context.today()),
+                                           list(inserted.get().wife().email(), inserted.get().husband().email(),
+                                                inserted.get().licence(), inserted.get().date())));
+                } else assert inserted.isEmpty();
+            });
+            tx.commit();
```
Comment (haikalpribadi) [inline simulation/agent/MarriageAgent.java]: This would be cleaner in terms of formatting (assuming you rename `partners` to `partner`).

```
            random.randomPairs(women, men).forEach(partner -> {
                String licence = partner.first().email() + partner.second().email();
                Optional<Marriage> inserted = insertMarriage(
                        tx, partner.first().email(), partner.second().email(), licence, context.today()
                );
                if (context.isReporting()) {
                    assert inserted.isPresent();
                    Marriage m = inserted.get();
                    reports.add(new Report(
                            list(partner.first().email(), partner.second().email(), licence, context.today()),
                            list(m.wife().email(), m.husband().email(), m.licence(), m.date())
                    ));
                } else assert inserted.isEmpty();
            });
```

File: simulation/agent/MarriageAgent.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        return context.seedData().countries();
+    }
+
+    @Override
+    protected List<Report> run(Session<TX> session, Country country, RandomSource random) {
+        List<Report> reports = new ArrayList<>();
+        try (TX tx = session.writeTransaction()) {
+            LocalDateTime partnerBirthDate = context.today().minusYears(context.ageOfAdulthood());
+            List<Person> women = matchPartner(tx, country, partnerBirthDate, Gender.FEMALE)
+                    .sorted(comparing(Person::email)).collect(toCollection(ArrayList::new));
```
Comment (haikalpribadi) [inline simulation/agent/MarriageAgent.java]: `.collect(toList());` would be shorter and cleaner.

File: simulation/agent/MarriageAgent.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+            List<Person> men = matchPartner(tx, country, partnerBirthDate, Gender.MALE)
+                    .sorted(comparing(Person::email)).collect(toCollection(ArrayList::new));
```
Comment (haikalpribadi) [inline simulation/agent/MarriageAgent.java]: `.collect(toList());` would be shorter and cleaner.

File: simulation/agent/ParenthoodAgent.java [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static com.vaticle.typedb.common.collection.Collections.list;
+import static java.util.Comparator.comparing;
+import static java.util.stream.Collectors.toCollection;
+
+public abstract class ParenthoodAgent<TX extends Transaction> extends Agent<Country, TX> {
```
Comment (haikalpribadi) [inline simulation/agent/ParenthoodAgent.java:41]: Why not just add parenthood relationship from the moment a child was born? Through `PersonAgent` (which may be renamed to `BirthAgent`? Sure, the first batch of births can't have parents.

File: typedb/simulation.tql [hunk 15]
```diff
@@ -42,15 +44,22 @@ person sub entity,
...
+    plays nationality:national,
+    plays citizenship:citizen,
     plays marriage:husband,
     plays marriage:wife,
-    plays parentship:parent,
-    plays parentship:child,
+    plays parenthood:parent,
+    plays parenthood:child,
+    plays grandparenthood:grandparent,
+    plays grandparenthood:grandchild,
```
Comment (haikalpribadi) [inline typedb/simulation.tql:57]: Why change `parentship` to `parenthood`? Doesn't "parentship" capture the relationship aspect of a parent and a child, as opposed to "parenthood" that captures only the state of the person being a parent?

File: typedb/simulation.tql [hunk 16]
```diff
@@ -42,15 +44,22 @@ person sub entity,
...
+    plays lineage:ancestor,
+    plays lineage:descendent,
     plays friendship:friend,
     plays employment:employee,
-    plays studentship:student,
+    plays study:student,
```
Comment (haikalpribadi) [inline typedb/simulation.tql:62]: Similar to "parentship", does "studentship" not capture the relationship more than than the word "study" that would often be interpreted as the act of studying?

File: typedb/simulation.tql [hunk 17]
```diff
@@ -105,15 +116,32 @@ residentship sub locates,
...
+
+parenthood sub relation,
     relates parent,
     relates child;
 
+grandparenthood sub relation,
+    relates grandparent,
+    relates grandchild;
+
+lineage sub relation,
```
Comment (haikalpribadi) [inline typedb/simulation.tql:141]: Is `ancestry` not a better word?

File: typedb/driver/TypeDBSession.java [hunk 18]
```diff
@@ -18,22 +18,34 @@
...
+import com.vaticle.typedb.client.api.connection.TypeDBOptions;
 
-import static com.vaticle.typedb.client.api.TypeDBTransaction.Type.WRITE;
+import static com.vaticle.typedb.client.api.connection.TypeDBTransaction.Type.READ;
+import static com.vaticle.typedb.client.api.connection.TypeDBTransaction.Type.WRITE;
 
 public class TypeDBSession implements Session<TypeDBTransaction> {
 
-    private final com.vaticle.typedb.client.api.TypeDBSession nativeSession;
+    private final com.vaticle.typedb.client.api.connection.TypeDBSession nativeSession;
```
Comment (haikalpribadi) [inline typedb/driver/TypeDBSession.java:28]: We've reverted this change in typedb-client-java. Let's upgrade and revert this change too.

File: typedb/agent/TypeDBNationalityAgent.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+import static com.vaticle.typedb.benchmark.typedb.Labels.COUNTRY;
+import static com.vaticle.typedb.benchmark.typedb.Labels.NATION;
+import static com.vaticle.typedb.benchmark.typedb.Labels.NATIONAL;
+import static com.vaticle.typedb.benchmark.typedb.Labels.NATIONALITY;
+import static com.vaticle.typedb.benchmark.typedb.Labels.PERSON;
+import static com.vaticle.typeql.lang.TypeQL.rel;
+import static com.vaticle.typeql.lang.TypeQL.var;
+
+public class TypeDBNationalityAgent extends NationalityAgent<TypeDBTransaction> {
+
```
Comment (haikalpribadi) [inline typedb/agent/TypeDBNationalityAgent.java:43]: remove extra line

File: typedb/agent/TypeDBLineageAgent.java [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+import static com.vaticle.typedb.benchmark.typedb.Labels.COUNTRY;
+import static com.vaticle.typedb.benchmark.typedb.Labels.DESCENDENT;
+import static com.vaticle.typedb.benchmark.typedb.Labels.LINEAGE;
+import static com.vaticle.typedb.benchmark.typedb.Labels.PERSON;
+import static com.vaticle.typedb.benchmark.typedb.Labels.PLACE;
+import static com.vaticle.typeql.lang.TypeQL.rel;
+import static com.vaticle.typeql.lang.TypeQL.var;
+
+public class TypeDBLineageAgent extends LineageAgent<TypeDBTransaction> {
+
```
Comment (haikalpribadi) [inline typedb/agent/TypeDBLineageAgent.java:50]: remove extra line

File: simulation/agent/ParenthoodAgent.java [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+                                                                 parenthood.second().email());
+                if (context.isReporting()) {
+                    assert inserted.isPresent();
+                    reports.add(new Report(list(parenthood.first().wife().email(), parenthood.first().husband().email(),
+                                                parenthood.second().email()),
+                                           list(inserted.get().mother().email(), inserted.get().father().email(),
+                                                inserted.get().child().email())));
+                } else assert inserted.isEmpty();
+            });
+            tx.commit();
```
Comment (haikalpribadi) [inline simulation/agent/ParenthoodAgent.java]: Formatting of this method can be made more readable by introducing variables in a few places.

File: neo4j/driver/Neo4jTransaction.java [hunk 22]
```diff
@@ -27,29 +27,57 @@
...
 
-public class Neo4jTransaction implements Transaction {
+public abstract class Neo4jTransaction implements Transaction {
 
     private static final Logger LOG = LoggerFactory.getLogger(Neo4jTransaction.class);
 
-    private final Session session;
+    protected final Session session;
 
     public Neo4jTransaction(Session session) {
```
Comment (haikalpribadi) [inline neo4j/driver/Neo4jTransaction.java:36]: Does this constructor need to remain public now it's always used internally?

File: neo4j/driver/Neo4jTransaction.java [hunk 23]
```diff
@@ -27,29 +27,57 @@
...
+        @Override
+        public void commit() {}
+    }
+
     public List<Record> execute(Query query) {
         return session.writeTransaction(tx -> {
             Result result = tx.run(query);
             return result.list();
         });
     }
```
Comment (haikalpribadi) [inline neo4j/driver/Neo4jTransaction.java]: Does this method still needs to exist now that the class is always extended?

**Review by haikalpribadi — CHANGES_REQUESTED** [review]
I've reviewed everything except for the TypeDB and Neo4j queries. Let's leave the queries and database model to last, after we fix all the current revisions. Also, next time, please remember to comment all the changes in the PR before submitting for revision.

File: Benchmark.java (hunk 3 above)
Comment (jmsfltchr) [reply Benchmark.java]: This is to fix an error I encountered when upgrading the `client-java` dependency:

`ERROR: /private/var/tmp/_bazel_jamesfletcher/9271413059d33570a857cdbbd3a205de/external/vaticle_typeql_lang_java/parser/BUILD:28:13: error loading package '@vaticle_typeql//grammar': in /private/var/tmp/_bazel_jamesfletcher/9271413059d33570a857cdbbd3a205de/external/vaticle_bazel_distribution/pip/rules.bzl: Unable to find package for @vaticle_bazel_distribution_pip//:requirements.bzl: The repository '@vaticle_bazel_distribution_pip' could not be resolved. and referenced by '@vaticle_typeql_lang_java//parser:parser'`

File: common/concept/Parenthood.java (hunk 5 above)
Comment (jmsfltchr) [reply common/concept/Parenthood.java:22]: So far this wasn't needed because the objects were never being compared. They should have been in the comparison test, but instead I was extracting the properties and comparing them. This can only have been from misreading the status-quo from the pre-existing agents; it's fixed now.

File: common/concept/Marriage.java (hunk 6 above)
Comment (jmsfltchr) [reply common/concept/Marriage.java:23]: Same reason and fix as above

File: common/params/Context.java (hunk 7 above)
Comment (jmsfltchr) [reply common/params/Context.java:112]: Given that this is totally arbitrary and will never affect the quantity or nature of the data generated, I can't see a reason why we would ever change it, can you? All of our other parameters have a real effect.

File: WORKSPACE (hunk 4 above)
Comment (jmsfltchr) [reply WORKSPACE:99]: This is to fix an error when the `client-java` dependency was updated. The error:
`
ERROR: /private/var/tmp/_bazel_jamesfletcher/9271413059d33570a857cdbbd3a205de/external/vaticle_typeql_lang_java/parser/BUILD:28:13: error loading package '@vaticle_typeql//grammar': in /private/var/tmp/_bazel_jamesfletcher/9271413059d33570a857cdbbd3a205de/external/vaticle_bazel_distribution/pip/rules.bzl: Unable to find package for @vaticle_bazel_distribution_pip//:requirements.bzl: The repository '@vaticle_bazel_distribution_pip' could not be resolved. and referenced by '@vaticle_typeql_lang_java//parser:parser'
`

File: simulation/agent/LineageAgent.java (hunk 8 above)
Comment (jmsfltchr) [reply simulation/agent/LineageAgent.java:38]: We've discussed this before and neither of us knew how to correct the generics. This issue arose after the refactor, and if I remember rightly addressing it the way we did prior to the refactor was overcomplicated.

File: simulation/agent/LineageAgent.java (hunk 9 above)
Comment (jmsfltchr) [reply simulation/agent/LineageAgent.java:59]: Lineages is transitive so will find an enormous number of relations. These dates scope the amount we are looking for by anchoring that the ancestors are the people from the first year of the simulation and the descendants were born in the current year. Perhaps though we would rather bind only one end of this transitive query since that's a common use-case.

File: simulation/agent/MarriageAgent.java (hunk 10 above)
Comment (jmsfltchr) [reply simulation/agent/MarriageAgent.java]: `partners` is grammatically correct but I agree doesn't read right, changing it to `pair`

File: simulation/agent/ParenthoodAgent.java (hunk 14 above)
Comment (jmsfltchr) [reply simulation/agent/ParenthoodAgent.java:41]: Surely that goes against the agent-based model? In terms of benchmarking also this would be a lot more work done in a single agent for readers to unpack. Let's discuss

File: typedb/simulation.tql (hunk 15 above)
Comment (jmsfltchr) [reply typedb/simulation.tql:57]: I did this purely on the basis that `parenthood` is an actual word that people commonly use, adding `ship` to the end is something we have normalised doing but I think it may confuse readers.

File: typedb/simulation.tql (hunk 16 above)
Comment (jmsfltchr) [reply typedb/simulation.tql:62]: Yes it does, but much more commonly studentship refers to a kind of scholarship though. Again, I don't want people confused so I changed it. `studentship` does have a second meaning in some dictionaries which is the one we want, so we can justify changing this back if we want.

File: typedb/simulation.tql (hunk 17 above)
Comment (jmsfltchr) [reply typedb/simulation.tql:141]: I deliberated this, decided against based on their definitions. The first definition of _ancestry_ is `one's family or ethnic descent` whereas _lineage_ is `direct descent from an ancestor; ancestry or pedigree`

File: typedb/driver/TypeDBSession.java (hunk 18 above)
Comment (jmsfltchr) [reply typedb/driver/TypeDBSession.java:28]: I've upgraded but I don't see this change reverted in `typedb-client-java`!

File: neo4j/driver/Neo4jTransaction.java (hunk 23 above)
Comment (jmsfltchr) [reply neo4j/driver/Neo4jTransaction.java]: I don't understand, this method is only present in the base interface `Transaction` and in the `Neo4jTransaction.Write` and `Neo4jTransaction.Read`, not in `Neo4jTransaction`. I've reordered the methods to be above the inner classes in `Neo4jTransaction` though as they were out of order.

File: Benchmark.java [hunk 24]
```diff
@@ -74,7 +74,9 @@ private static FactoryTracing initTracing(@Nullable Options.FactoryTracing optio
             tracing = FactoryTracing.create(options.factory(), cred.username(), cred.token()).withLogging();
         }
         FactoryTracingThreadStatic.setGlobalTracingClient(tracing);
-        FactoryTracingThreadStatic.openGlobalAnalysis(options.org(), options.repo(), options.commit(), analysisName);
+        String taggedAnalysisName = analysisName;
```
Comment (haikalpribadi) [inline Benchmark.java:77]: Why do we need to rename `analysisName` to `taggedAnalysisName`? Doesn't `analysisName` as the variable name to be used throughout?

File: neo4j/driver/Neo4jTransaction.java (hunk 22 above)
Comment (haikalpribadi) [reply neo4j/driver/Neo4jTransaction.java:36]: I think you may have forgotten about this comment.

File: simulation/agent/ParenthoodAgent.java (hunk 14 above)
Comment (haikalpribadi) [reply simulation/agent/ParenthoodAgent.java:41]: The work in the one agent does get more involved, I agree. But I suppose it's an "atomic" event in life? Yes, let's discuss in person.

File: typedb/simulation.tql (hunk 15 above)
Comment (haikalpribadi) [reply typedb/simulation.tql:57]: Oh really? I thought `parentship` is a real word too! 🤣 https://www.lexico.com/definition/parentship

File: typedb/simulation.tql (hunk 16 above)
Comment (haikalpribadi) [reply typedb/simulation.tql:62]: Interesting! I didn't immediately pick up that `studentship` could mean `scholarship`. Now that you point it out I can see it too. However, this may be because I'm missing the colloquial knowledge/context, but `study` also feels confusing to me as it doesn't describe a "relationship" at all to me. "Study" just sounds like the of studying.

File: typedb/simulation.tql (hunk 17 above)
Comment (haikalpribadi) [reply typedb/simulation.tql:141]: I see. But I think "lineage" as another (more commonly used) connotation, which is: the "data lineage". I worry that "lineage" in our industry would be confused too easily to metadata traceability.

**Review by haikalpribadi — CHANGES_REQUESTED** [review]
Just a few more points to take action on, the rest we can discuss.

File: typedb/simulation.tql (hunk 17 above)
Comment (jmsfltchr) [reply typedb/simulation.tql:141]: I'm definitely not concerned about that given the whole scope of the schema is clearly about people, and the roles of the relation tell you that already.

File: neo4j/driver/Neo4jTransaction.java (hunk 22 above)
Comment (jmsfltchr) [reply neo4j/driver/Neo4jTransaction.java:36]: Oops missed it, addressed now

File: Benchmark.java (hunk 24 above)
Comment (jmsfltchr) [reply Benchmark.java:77]: This is to avoid having a side-effect on the parameter `analysisName` that was passed in but reading into it I realise that this is unnecessary.

**Review by haikalpribadi — APPROVED** [review]
